src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -987,15 +987,15 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
         if (ASRUtils::is_integer(*left_type) && ASRUtils::is_integer(*right_type)) {
             bool is_l64 = ASR::down_cast<ASR::Integer_t>(left_type)->m_kind == 8;
             bool is_r64 = ASR::down_cast<ASR::Integer_t>(right_type)->m_kind == 8;
-            if (is_l64 != is_r64) {
+            if (is_l64 && !is_r64) {
```

Comment:
I think this breaks binary operations like `a + b` if right is 64 bit and left is 32 bit.

In binary operations, the cast has to happen from 32 bit to 64 bit no matter if it is LHS or RHS that needs casting.

Reply:
I think it won't be an issue as it will be handled by:
https://github.com/lcompilers/lpython/blob/b3154087db644b951b6fef745408c3a875a9a207/src/lpython/semantics/python_ast_to_asr.cpp#L1379-L1383

Reply:
ok. Let me carefully review this to make sure we don't introduce an error here.

Reply:
It can be verified from the following example:
```py
def main0():
    i: i64
    j: i32
    i = 0
    j = 1
    print(i + j)
    print(j + i)

main0()
```

<details>
<summary> ASR </summary>
<p>

```bash
(TranslationUnit 
   (SymbolTable 
      1 
      {
         _lpython_main_program: 
            (Subroutine 
               (SymbolTable 
                  4 
                  {
                  }) 
               _lpython_main_program [] [
               (SubroutineCall 1 main0 () [] ())] 
               Source 
               Public 
               Implementation () .false. .false.), 
         main0: 
            (Subroutine 
               (SymbolTable 
                  2 
                  {
                     i: 
                        (Variable 
                           2 
                           i 
                           Local () () 
                           Default 
                           (Integer 8 []) 
                           Source 
                           Public 
                           Required .false.), 
                     j: 
                        (Variable 
                           2 
                           j 
                           Local () () 
                           Default 
                           (Integer 4 []) 
                           Source 
                           Public 
                           Required .false.)
                  }) 
               main0 [] [
               (= 
                  (Var 2 i) 
                  (ImplicitCast (ConstantInteger 0 
                     (Integer 4 [])) 
                     IntegerToInteger 
                     (Integer 8 []) ()) ()) 
               (= 
                  (Var 2 j) (ConstantInteger 1 
                  (Integer 4 [])) ()) 
               (Print () [
                  (BinOp 
                     (Var 2 i) 
                     Add 
                     (ImplicitCast 
                        (Var 2 j) 
                        IntegerToInteger 
                        (Integer 8 []) ()) 
                     (Integer 8 []) () ())]) 
               (Print () [
                  (BinOp 
                     (ImplicitCast 
                        (Var 2 j) 
                        IntegerToInteger 
                        (Integer 8 []) ()) 
                     Add 
                     (Var 2 i) 
                     (Integer 8 []) () ())])] 
               Source 
               Public 
               Implementation () .false. .false.), 
         main_program: 
            (Program 
               (SymbolTable 
                  3 
                  {
                  }) 
               main_program [] [
               (SubroutineCall 1 _lpython_main_program () [] ())])
      }) 
   [])
```

</p>
</details>